analyzer/src/steps/resolve.rs

```diff
@@ -76,46 +103,37 @@ impl<'a, 'e> SymbolResolver<'a, 'e> {
     /// enables the resolution and pushes diagnostics if any symbol could not be resolved.
     fn resolve(&mut self) {
         let mut unresolved_imports = self.relations.take_imports();
-
+        let mut resolved_imports = HashMap::new();
         for (env_id, env) in self.engine.environments() {
             let unresolved_imports = unresolved_imports.remove(&env_id).unwrap_or_default();
 
-            let resolved_imports = self.resolve_imports(env_id, unresolved_imports);
-            self.resolve_symbols_of(env_id, env, resolved_imports);
+            let local_resolved_imports = self.resolve_imports(env_id, unresolved_imports);
+            self.resolve_symbols_of(env, &local_resolved_imports);
+            resolved_imports.insert(env_id, local_resolved_imports);
         }
+        self.resolve_trees(&resolved_imports);
     }
 
     /// resolves the symbols of given environment, using given resolved imports for external symbol references.
-    fn resolve_symbols_of(
-        &mut self,
-        env_id: SourceObjectId,
-        env: &Environment,
-        imports: ResolvedImports,
-    ) {
+    fn resolve_symbols_of(&mut self, env: &Environment, imports: &ResolvedImports) {
         for (symbol_name, external_var) in env.variables.external_vars() {
             let object = &mut self.relations.objects[external_var.0];
 
-            match imports.imported_symbols.get(symbol_name) {
-                Some(resolved_symbol) => object.resolved = Some(*resolved_symbol),
-                None => self.diagnose_unresolved_external_symbols(
-                    external_var,
-                    env_id,
-                    env,
-                    symbol_name,
-                ),
+            if let Some(resolved_symbol) = imports.imported_symbols.get(symbol_name) {
+                object.resolved = Some(*resolved_symbol);
             };
         }
     }
 
     /// Appends a diagnostic for an external symbol that could not be resolved.
-    /// Each expression that used this symbol (such as VarReferences) will then get an observation
+    ///
+    /// Each expression that use this symbol (such as variable references) will then get an observation.
     fn diagnose_unresolved_external_symbols(
-        &mut self,
         external_var: GlobalObjectId,
         env_id: SourceObjectId,
         env: &Environment,
         name: &str,
-    ) {
+    ) -> Diagnostic {
```

Comment:
If you change this, i think you should also make the same change for `diagnose_unresolved_import`